File: haskell/task2-v4.hs
{-Знайти всі “нижні піки” списку та підрахувати їх кількість. Елемент списку є нижнім піком, якщо він менший
за своїх наявних сусідів. Вважати, що список складається з різних елементів. Наприклад, у списку
[5,4,2,8,3,1,6,9,5] нижні піки та їх позиції такі: (2,3), (1,6), (5,9).-}
lowerPeaks :: [Int] -> ([(Int, Int)], Int)
lowerPeaks xs =
    let peaks = getPeaks Nothing xs 1
    in (peaks, countList peaks)
  where
    getPeaks :: Maybe Int -> [Int] -> Int -> [(Int, Int)]
    getPeaks _ [] _ = []
    getPeaks Nothing [_] _ = []
    
    -- останній елемент
    getPeaks (Just left) [x] pos =
        [(x, pos) | x < left]
        
    -- перший елемент
    getPeaks Nothing (x:y:xs) pos =
        let rest = getPeaks (Just x) (y:xs) (pos + 1)
        in if x < y
           then (x, pos) : rest
           else rest
           
    -- інші
    getPeaks (Just left) (x:y:xs) pos =
        let rest = getPeaks (Just x) (y:xs) (pos + 1)
        in if x < left && x < y
           then (x, pos) : rest
           else rest
    countList :: [a] -> Int
    countList [] = 0
    countList (_:ts) = 1 + countList ts

main :: IO ()
main = do
    putStrLn "--- Test 1 ---"
    putStrLn "Input: [5, 4, 2, 8, 3, 1, 6, 9, 5]"
    putStrLn "Expected: ([(2,3),(1,6),(5,9)],3)"
    putStrLn "Result:"
    print (lowerPeaks [5, 4, 2, 8, 3, 1, 6, 9, 5])

    putStrLn "\n--- Test 2 ---"
    putStrLn "Input: [10, 8, 6, 4]"
    putStrLn "Expected: ([(4,4)],1)"
    putStrLn "Result:"
    print (lowerPeaks [10, 8, 6, 4])

    putStrLn "\n--- Test 3  ---"
    putStrLn "Input: [2, 5, 8, 12]"
    putStrLn "Expected: ([(2,1)],1)"
    putStrLn "Result:"
    print (lowerPeaks [2, 5, 8, 12])

    putStrLn "\n--- Test 4 ---"
    putStrLn "Input: []"
    putStrLn "Expected: ([],0)"
    putStrLn "Result:"
    print (lowerPeaks [])

    putStrLn "\n--- Test 5 ---"
    putStrLn "Input: [1]"
    putStrLn "Expected: ([],0)"
    putStrLn "Result:"
    print (lowerPeaks [1])

    putStrLn "\n--- Test 6 ---"
    putStrLn "Input: [7, 3]"
    putStrLn "Expected: ([(3,2)],1)"
    putStrLn "Result:"
    print (lowerPeaks [7, 3])

    putStrLn "\n--- Test 7 ---"
    putStrLn "Input: [3, 8, 10, 6, 2]"
    putStrLn "Expected: ([(3,1), (2,5)],2)"
    putStrLn "Result:"
    print (lowerPeaks [3, 8, 10, 6, 2])

{-
main :: IO ()
main = do
    putStrLn "Enter list of different numbers (e.g. 5 4 2 8 3 1 6 9 5):"
    input <- getLine
    
    let numbers = map read (words input) :: [Int]
    
    putStrLn "\nResult (Peaks, Count):"
    print (lowerPeaks numbers)
-}

File: prolog/task2-v4.pl
lower_peaks(List, (Peaks, Count)) :-
    find_peaks(List, Peaks),
    length(Peaks, Count).

find_peaks([], []).
find_peaks([_], []).

find_peaks([First, Second | Rest], Peaks) :-
    (
        First < Second ->
            Peaks = [(First, 1) | OtherPeaks]
        ;
            Peaks = OtherPeaks
    ),
    find_peaks_middle(First, Second, Rest, 2, OtherPeaks).

find_peaks_middle(Left, Current, [], Pos, Peaks) :-
    (
        Current < Left ->
            Peaks = [(Current, Pos)]
        ;
            Peaks = []
    ).

find_peaks_middle(Left, Current, [Right | Rest], Pos, Peaks) :-
    NextPos is Pos + 1,
    (
        Current < Left,
        Current < Right ->
            Peaks = [(Current, Pos) | OtherPeaks]
        ;
            Peaks = OtherPeaks
    ),
    find_peaks_middle(Current, Right, Rest, NextPos, OtherPeaks).

run_tests :-
    writeln('--- Test 1 ---'),
    writeln('Input: [5, 4, 2, 8, 3, 1, 6, 9, 5]'),
    writeln('Expected: ([(2,3),(1,6),(5,9)],3)'),
    writeln('Result:'),
    lower_peaks([5, 4, 2, 8, 3, 1, 6, 9, 5], Result1),
    writeln(Result1),

    nl,
    writeln('--- Test 2 ---'),
    writeln('Input: [10, 8, 6, 4]'),
    writeln('Expected: ([(4,4)],1)'),
    writeln('Result:'),
    lower_peaks([10, 8, 6, 4], Result2),
    writeln(Result2),

    nl,
    writeln('--- Test 3 ---'),
    writeln('Input: [2, 5, 8, 12]'),
    writeln('Expected: ([(2,1)],1)'),
    writeln('Result:'),
    lower_peaks([2, 5, 8, 12], Result3),
    writeln(Result3),

    nl,
    writeln('--- Test 4 ---'),
    writeln('Input: []'),
    writeln('Expected: ([],0)'),
    writeln('Result:'),
    lower_peaks([], Result4),
    writeln(Result4),

    nl,
    writeln('--- Test 5 ---'),
    writeln('Input: [1]'),
    writeln('Expected: ([],0)'),
    writeln('Result:'),
    lower_peaks([1], Result5),
    writeln(Result5),

    nl,
    writeln('--- Test 6 ---'),
    writeln('Input: [7, 3]'),
    writeln('Expected: ([(3,2)],1)'),
    writeln('Result:'),
    lower_peaks([7, 3], Result6),
    writeln(Result6),

    nl,
    writeln('--- Test 7 ---'),
    writeln('Input: [3, 8, 10, 6, 2]'),
    writeln('Expected: ([(3,1),(2,5)],2)'),
    writeln('Result:'),
    lower_peaks([3, 8, 10, 6, 2], Result7),
    writeln(Result7).